encoding/codecv3/codecv3.go

```diff
@@ -0,0 +1,243 @@
+
+	"github.com/scroll-tech/da-codec/encoding"
+	"github.com/scroll-tech/da-codec/encoding/codecv1"
+	"github.com/scroll-tech/da-codec/encoding/codecv2"
+)
+
+// DABlock represents a Data Availability Block.
+type DABlock = codecv2.DABlock
+
+// DAChunk groups consecutive DABlocks with their transactions.
+type DAChunk = codecv2.DAChunk
+
+// DABatch contains metadata about a batch of DAChunks.
+type DABatch struct {
+	// header
+	Version                uint8
+	BatchIndex             uint64
+	L1MessagePopped        uint64
+	TotalL1MessagePopped   uint64
+	DataHash               common.Hash
+	BlobVersionedHash      common.Hash
+	ParentBatchHash        common.Hash
+	LastBlockTimestamp     uint64
+	BlobDataProof          [64]byte
+	SkippedL1MessageBitmap []byte
+
+	// blob payload
+	blob *kzg4844.Blob
+	z    *kzg4844.Point
+}
+
+// NewDABlock creates a new DABlock from the given encoding.Block and the total number of L1 messages popped before.
+func NewDABlock(block *encoding.Block, totalL1MessagePoppedBefore uint64) (*DABlock, error) {
+	return codecv2.NewDABlock(block, totalL1MessagePoppedBefore)
+}
+
+// NewDAChunk creates a new DAChunk from the given encoding.Chunk and the total number of L1 messages popped before.
+func NewDAChunk(chunk *encoding.Chunk, totalL1MessagePoppedBefore uint64) (*DAChunk, error) {
+	return codecv2.NewDAChunk(chunk, totalL1MessagePoppedBefore)
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch) (*DABatch, error) {
+	// this encoding can only support a fixed number of chunks per batch
+	if len(batch.Chunks) > codecv2.MaxNumChunks {
+		return nil, fmt.Errorf("too many chunks in batch")
+	}
+
+	if len(batch.Chunks) == 0 {
+		return nil, fmt.Errorf("too few chunks in batch")
+	}
+
+	if len(batch.Chunks[len(batch.Chunks)-1].Blocks) == 0 {
+		return nil, fmt.Errorf("too few blocks in last chunk of the batch")
+	}
+
+	// batch data hash
+	dataHash, err := codecv1.ComputeBatchDataHash(batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// skipped L1 messages bitmap
+	bitmapBytes, totalL1MessagePoppedAfter, err := encoding.ConstructSkippedBitmap(batch.Index, batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// blob payload
+	blob, blobVersionedHash, z, err := codecv2.ConstructBlobPayload(batch.Chunks, false /* no mock */)
+	if err != nil {
+		return nil, err
+	}
+
+	daBatch := DABatch{
+		Version:                uint8(encoding.CodecV3),
+		BatchIndex:             batch.Index,
+		L1MessagePopped:        totalL1MessagePoppedAfter - batch.TotalL1MessagePoppedBefore,
+		TotalL1MessagePopped:   totalL1MessagePoppedAfter,
+		DataHash:               dataHash,
+		BlobVersionedHash:      blobVersionedHash,
+		ParentBatchHash:        batch.ParentBatchHash,
+		LastBlockTimestamp:     batch.Chunks[len(batch.Chunks)-1].Blocks[len(batch.Chunks[len(batch.Chunks)-1].Blocks)-1].Header.Time,
```

Comment:
it's too long. How about changing to an individual variable. And I wonder the array bound if don't check the boundry

Reply:
sounds good. and the array length (boundary) is checked above in the same function.

Reply:
changed in [6b8fd1e](https://github.com/scroll-tech/da-codec/pull/11/commits/6b8fd1e6996c4dd1c59a50adc0af57c1dfc279ae).